### PR TITLE
Clean shared project directory when disk usage is too high

### PR DESCRIPTION
**Pull request by chengren311**
This PR implements LRU project purging to prevent project files eating up disk space on executor.

We will encapsulate following logic inside the background cleanup thread:
if(disk space consumed by projects >= predefined threshold) 
   List all project files and sort them by creation time in ascending order
   Iterate over the project file list, delete the file if the project is not running until disk size of projects drops down below a predefined lower threshold or no more projects to delete.

Why using creation time not last access time:
Directory last access time is not maintained by most file systems so we cannot rely on file system API to get last access time. An alternative considered is to modify the project dir every time the code reads it, then last modification time would be equivalent of last access time. The associate cons are code complexity and overhead of disk IO. We will use last creation time as the indicator of oldness of the project. Although creation time might not be as indicative as last access time to determine the hotness of a project, we are still ok with using it based on the assumption that the older the project is, more likely it’s not being used.

This PR also changes execution dir retention from 1 day to 2 hours. Since execution dir is hard linked to project dir, so disk space will be released only when no reference to project dir exists. That's why we want to shorten the execution dir retention time so that disk pressure can be alleviated sooner.

Project cleanup is executed by active executor only to prevent ensure only one thread will be performing deletion. Currently it relies on FlowRunnerManager#isExecutorActive to determine whether itself is active. But isExecutorActive is reliable only when azkaban admin uses activate API to active the executor. If admin manually updates the executors table in database and call reloadExecutors API, then the flag won't be set which inactive executor to perform deletion. So improvement need to be done here to make sure isExecutorActive is authentic.

### CODE REVIEW

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java [hunk 1]
```diff
@@ -69,6 +73,32 @@ public static boolean isDirWritable(final File dir) {
...
+    final BasicFileAttributes attrs = Files
+        .readAttributes(file.toPath(), BasicFileAttributes.class);
+    final FileTime time = attrs.creationTime();
+    return time.toMillis();
+  }
+
+  /**
+   * Return the size of dir in KB.
+   */
+  public static long sizeInKB(final File dir) throws IOException {
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: Can we use `FileUtils.sizeOfDirectory` to get the size directly?

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java [hunk 2]
```diff
@@ -69,6 +73,32 @@ public static boolean isDirWritable(final File dir) {
     return true;
   }
 
+
+  /**
+   * Return creation time of file.
+   */
+  public static long creationTime(final File file) throws IOException {
```
Comment (jamiesjc) [inline azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: Does `getCreationTime` sound better?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java [hunk 3]
```diff
@@ -70,14 +70,16 @@ void setup(final ExecutableFlow flow) {
       // First get the ProjectVersion
       final ProjectVersion projectVersion = getProjectVersion(flow);
 
-      // Setup the project
-      setupProject(projectVersion);
+      synchronized (projectVersion) {
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java]: What's the intention of synchronizing on projectVersion here?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 4]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
       this.projectDirectory.mkdirs();
     }
 
+    this.projectDirMaxSizeInMB = props.getLong(ConfigurationKeys.PROJECT_DIR_MAX_SIZE,
+        2000000); // default value as 2TB
+    this.projectDirStartDeletionThreshold = props
+        .getDouble(ConfigurationKeys.PROJECT_DIR_CLEANUP_START_THRESHOLD, 0.9);
+    this.projectDirStopDeletionThreshold = props
+        .getDouble(ConfigurationKeys.PROJECT_DIR_CLEANUP_STOP_THRESHOLD, 0.8);
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Is there any reason why we set the threshold between 0.8-0.9?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 5]
```diff
@@ -970,6 +992,102 @@ private void cleanOlderProjects() {
...
+          new HashSet<>();
+      for (final FlowRunner runner : FlowRunnerManager.this.runningFlows.values()) {
+        final ExecutableFlow flow = runner.getExecutableFlow();
+        activeProjectVersions.add(new Pair<>(flow
+            .getProjectId(), flow.getVersion()));
+      }
+      return activeProjectVersions;
+    }
+
+    private boolean isActiveProject(final ProjectVersion version) {
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Do we define active projects as the projects that have currently running flows? What if they have scheduled flows? Should it also be considered as active?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 6]
```diff
@@ -970,6 +992,102 @@ private void cleanOlderProjects() {
...
+            } else {
+              return 0;
+            }
+          } catch (final IOException ex) {
+            logger.error("error reading project dir", ex);
+            return 0;
+          }
+        });
+
+        for (final ProjectVersion version : projectVersionList) {
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Will it take a long time to calculate the project cache directory size since it's very big? And it will be called inside `shouldKeepCleanup` method which gets executed for every project version in the for loop. Would it be more efficient to do some batch cleaning?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 4 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: it's a reasonable number.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 5 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Active projects are the one containing currently running flows.

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java (hunk 1 above)
Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: sizeOfDirectory is not precise to get the read disk usage. E.g directory itself has disk usage but it doesn't count in sizeOfDirectory.

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java (hunk 2 above)
Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: updated

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java (hunk 3 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java]: updated in comment

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 6 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: yes, but this the frequency of doing it is once/30 mins. 
What do you mean by batch cleaning?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java [hunk 7]
```diff
@@ -70,17 +70,22 @@ void setup(final ExecutableFlow flow) {
...
-
-      // Create the execution directory
-      execDir = createExecDir(flow);
-
-      // Create the symlinks from the project
-      copyCreateHardlinkDirectory(projectVersion.getInstalledDir(), execDir);
+      // Synchronized on {@code projectVersion} to prevent
+      // cleaning thread in {@link azkaban.execapp.FlowRunnerManager}
+      // cleaning up the same project version when creating hardlink
+      synchronized (projectVersion) {
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java]: In `getProjectVersion()`, it seems the projectVersion returned is a new object. Will the synchronization really work if they are synchronized on different objects here?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 6 above)
Comment (jamiesjc) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: I was thinking if it might be possible to clean up like 10 projects at a time and then do the check in `shouldKeepCleanup`. But I'm not sure how much improvement this will have depending on how long it will take in calculating the directory size. If you think efficiency should not be a problem, then I'm ok with this implementation.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 8]
```diff
@@ -134,7 +142,7 @@
   private Props globalProps;
 
   private long lastCleanerThreadCheckTime = -1;
-  private long executionDirRetention = 1 * 24 * 60 * 60 * 1000; // 1 Day
+  private long executionDirRetention = 1 * 6 * 60 * 60 * 1000; // 6 hours
```
Comment (jamiesjc) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Why change it to 6 hours?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java (hunk 7 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java]: it will only return new object only when projectId, version don't present in the hash map.
```   
 final ProjectVersion projectVersion;
    synchronized (this.installedProjects) {
      projectVersion = this.installedProjects
          .computeIfAbsent(new Pair<>(flow.getProjectId(), flow.getVersion()),
              k -> new ProjectVersion(flow.getProjectId(), flow.getVersion()));
    }
    return projectVersion;
 ```

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 8 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: reduce it to 2 hours.
```
This PR also changes execution dir retention from 1 day to 2 hours. Since execution dir is hard linked to project dir, so disk space will be released only when no reference to project dir exists. That's why we want to shorten the execution dir retention time so that disk pressure can be alleviated sooner.

```

File: az-core/src/main/java/azkaban/Constants.java [hunk 9]
```diff
@@ -224,6 +224,21 @@
     public static final String QUEUEPROCESSING_ENABLED = "azkaban.queueprocessing.enabled";
 
     public static final String SESSION_TIME_TO_LIVE = "session.time.to.live";
+
+    // allowed max size of project dir in mb
+    public static final String PROJECT_DIR_MAX_SIZE = "azkaban.project.dir_max_size";
```
Comment (HappyRay) [inline az-core/src/main/java/azkaban/Constants.java]: Consider adding the unit in the key name? e.g.
dir_max_size_in_mb?

Do you foresee other keys in the project namespace? Project-dir can be interpreted as an individual project directory. 
Would it be better to name it 
azkaban.project_cache_max_size_in_mb?

File: az-core/src/main/java/azkaban/Constants.java [hunk 10]
```diff
@@ -224,6 +224,21 @@
...
+
+    // the disk usage threshold to trigger project dir cleanup.
+    // E.g, if set to 0.8, cleanup will trigger when project dir size >= 0.8 of
+    // {@value#PROJECT_DIR_MAX_SIZE}.
+    public static final String PROJECT_DIR_CLEANUP_START_THRESHOLD = "azkaban.project"
+        + ".start_cleanup_threshold";
+
+    // the disk usage threshold to stop project dir cleanup.
+    // E.g, if set to 0.6, cleanup will stop when project dir size < 0.6 of
```
Comment (HappyRay) [inline az-core/src/main/java/azkaban/Constants.java]: What do you think about making this an integer?
e.g.
60 means 60%?

**Comment by HappyRay** [comment]
Where are the tests for this change?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 11]
```diff
@@ -134,7 +142,7 @@
   private Props globalProps;
 
   private long lastCleanerThreadCheckTime = -1;
-  private long executionDirRetention = 1 * 24 * 60 * 60 * 1000; // 1 Day
+  private long executionDirRetention = 1 * 2 * 60 * 60 * 1000; // 2 hours
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Is this change necessary to be included in this change?

This is not related to the project cache directory.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 12]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
       this.projectDirectory.mkdirs();
     }
 
+    this.projectDirMaxSizeInMB = props.getLong(ConfigurationKeys.PROJECT_DIR_MAX_SIZE,
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: "projectDirMaxSizeInMb" see coding standard naming.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 13]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
...
+        2000000); // default value as 2TB
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Why is 2 TB a good default? I don't think most of the deployment will require such a big cache.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 14]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
...
+    this.projectDirStartDeletionThreshold = props
+        .getDouble(ConfigurationKeys.PROJECT_DIR_CLEANUP_START_THRESHOLD, 0.9);
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Is 10% buffer big enough when the default max size is smaller?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 15]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
...
+    this.projectDirStopDeletionThreshold = props
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: I would suggest a lower number so that each clean up can clean up at least 20% or more of the max to reduce the number of times the clean up needs to run and thus reducing the amortized cost of each clean up.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 16]
```diff
@@ -173,6 +181,16 @@ public FlowRunnerManager(final Props props,
...
+        .getDouble(ConfigurationKeys.PROJECT_DIR_CLEANUP_STOP_THRESHOLD, 0.8);
+    Preconditions.checkArgument(this.projectDirStartDeletionThreshold >= 0 && this
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: These are not arguments. It seems a bit odd.

Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Consider adding methods to get config values and add validation logic there.

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 17]
```diff
@@ -781,11 +799,14 @@ public void deleteExecutionDirectory() {
     private static final long OLD_PROJECT_DIR_INTERVAL_MS = 5 * 60 * 1000;
     // Every 2 mins clean the recently finished list
     private static final long RECENTLY_FINISHED_INTERVAL_MS = 2 * 60 * 1000;
+    // Every 30 mins clean least recently used project dir
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Should this value be configurable?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 18]
```diff
@@ -840,6 +861,13 @@ public void run() {
               this.lastExecutionDirCleanTime = currentTime;
             }
 
+            if (currentTime - LRU_PROJECT_DIR_INTERVAL_MS > this.lastLRUProjectCleanTime
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Make this block a private method?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java [hunk 19]
```diff
@@ -970,6 +992,105 @@ private void cleanOlderProjects() {
         }
       }
     }
+
+    private HashSet<Pair<Integer, Integer>> getActiveProjectVersions() {
+      final HashSet<Pair<Integer, Integer>> activeProjectVersions =
+          new HashSet<>();
+      for (final FlowRunner runner : FlowRunnerManager.this.runningFlows.values()) {
+        final ExecutableFlow flow = runner.getExecutableFlow();
```
Comment (HappyRay) [inline azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: What happens if the runningFlows changes between this statement and the call to .values()?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 6 above)
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: Have you measured the performance of the current design?

In case performance is proven to be an issue, have you considered the following alternative design? 

Delete older versions of the project immediately. This will free up disk space more quickly.

Create a file in each project directory and write the size of the project to the file when the project is created. The project files are not supposed to change after creation. Touch this file each time the project is used. This way, we can have a more efficient LRU algorithm based on last access time, not creation time. 

Maintain the total size of the project cache in memory to avoid the overhead of re-calculating it. The size shouldn't change too often. This way we can afford to run the check more frequently.

**Comment by HappyRay** [comment]
How do you plan to monitor this feature?

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java (hunk 1 above)
Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: high accuracy doesn't hurt, also FileUtils.sizeOfDirectory doesn't really indicate the real disk usage of the directory as size of a file is not equivalent of its associated disk cost. https://superuser.com/questions/94217/why-ls-and-du-show-different-size

File: az-core/src/main/java/azkaban/Constants.java [hunk 20]
```diff
@@ -224,6 +224,21 @@
...
+    // {@value#PROJECT_DIR_MAX_SIZE}.
+    public static final String PROJECT_DIR_CLEANUP_STOP_THRESHOLD = "azkaban.project"
+        + ".stop_cleanup_threshold";
```
[the remark this answers is not in the document]
Comment (chengren311) [reply az-core/src/main/java/azkaban/Constants.java]: it's auto formatted by save plugin.

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java (hunk 1 above)
Comment (HappyRay) [reply azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: The downside is unnecessary complexity. 

The referenced article seems to refer to a sparse file. The project files are not sparse files, are they?

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java (hunk 7 above)
Comment (HappyRay) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowPreparer.java]: Was it subject to race conditions then? Could you describe the problem you tried to fix in more details?

File: az-core/src/main/java/azkaban/Constants.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (HappyRay) [reply az-core/src/main/java/azkaban/Constants.java]: public static final String PROJECT_DIR_CLEANUP_START_THRESHOLD =
        "azkaban.project.start_cleanup_threshold";
seems to work for me.

File: az-core/src/main/java/azkaban/Constants.java (hunk 9 above)
Comment (chengren311) [reply az-core/src/main/java/azkaban/Constants.java]: updated  in https://github.com/azkaban/azkaban/pull/1812

File: az-core/src/main/java/azkaban/Constants.java (hunk 10 above)
Comment (chengren311) [reply az-core/src/main/java/azkaban/Constants.java]: updated in https://github.com/azkaban/azkaban/pull/1812

File: az-core/src/main/java/azkaban/Constants.java (hunk 20 above)
[the remark this answers is not in the document]
Comment (chengren311) [reply az-core/src/main/java/azkaban/Constants.java]: updated in #1812

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 12 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: updated in #1812

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 13 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: updated in #1812

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 14 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: updated in #1812

File: azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java (hunk 15 above)
Comment (chengren311) [reply azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java]: updated in #1812

File: azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java (hunk 1 above)
Comment (chengren311) [reply azkaban-common/src/main/java/azkaban/utils/FileIOUtils.java]: FileUtils.sizeOfDirectory is a lot slower than running du -sh based on my experiment of reading 10000+ project dir.
